src/protocol/sort/apply.rs

```diff
@@ -0,0 +1,103 @@
+use embed_doc_image::embed_doc_image;
+/// Permutation reorders (1, 2, . . . , m) into (σ(1), σ(2), . . . , σ(m)).
+/// For example, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is reordered into (C, D, B, A) by σ.
+#[embed_doc_image("apply", "images/sort/apply.png")]
+#[allow(dead_code)]
+pub fn apply<T: Copy + Default>(source_indices: &[usize], values: &[T]) -> Vec<T> {
+    let mut applied_vec: Vec<T> = (0..source_indices.len())
+        .map(|_| Default::default())
+        .collect();
+    (0..source_indices.len()).for_each(|i| {
+        let idx = source_indices[i];
+        applied_vec[i] = values[idx];
+    });
```

Comment:
Can this be combined into just one loop? This appears to be looping over the permutation just to assign a default value, then looping a second time to assign things.

Reply:
Replacing this function by a rust crate permutation : https://docs.rs/permutation/latest/permutation/permutation/struct.Permutation.html
it has functions to apply permutation and inverse in place. However it uses cycle decomposition (hence can be open to a potential timely attack). Confirming on that with Ryo.